### PR TITLE
SIG Addons: Procedure and RFC template for Addons migrations

### LINKED ISSUE #239: Tensorflow addons promotion and TF
Please take an overview of https://github.com/tensorflow/tensorflow/pull/33945#issuecomment-621117201

Can we document somewhere that new ops PR on Tensorflow duplicating TF addons implementations can pass through a small RFC proposal here (so that we include Addons removal or migration in the proposal)?

Currently we have https://github.com/tensorflow/addons/pull/774 that was approved directly inside the SIG but here we don't have any process documented on the TF Team side.

That PR was derived from a ticket discussed with @karmel @ewilderj  at https://github.com/tensorflow/addons/issues/686

### PR DESCRIPTION
**Pull request by seanpmorgan**
Tried to incorporate the discussion points in #239 into documentation. Closes #239 once agreed upon.

The goal of this PR is to standardize how Addons components will be migrated going forward. After the RFC template and and Procedure are agreed upon we will submit the proposal for GELU migration.

cc @karmel @alextp @bhack  for feedback

### CODE REVIEW

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Migration From TF-Addons To TensorFlow Core
+
+### In-Progress & Previous Migrations:
+https://github.com/tensorflow/addons/projects/2/
```
Comment (seanpmorgan) [inline sigs/addons/MIGRATION_TO_CORE.md:4]: @theadactyl This may be more visible to a larger audience if the project is housed in tf/community

File: sigs/addons/MIGRATION_RFC_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+## Changes to Implementation (If Needed)
+```
+Code snippet(s) showing proposed code changes
+```
+* Discussion on the rationale for changes
+
+## Transition Plan
+* Proposed landing place in tf-core
+* Deprecation plan for Addons
+    * Will we be able to alias the core implementation? (parameters must be exact match)
```
Comment (bhack) [inline sigs/addons/MIGRATION_RFC_TEMPLATE.md:34]: I think that the upstreamed version could, under a review request, acquire a parameter superset. But if the additional paramas are optional I think that it is still ok if we have the same output with the parameter subset.

File: sigs/addons/MIGRATION_RFC_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+| **Authors** | XXXXXXXXXX   |
+| **Sponsor** | XXXXXXXXXX   |
+| **Updated** | YYYY-MM-DD   |
+| **Sponsorship Deadline** | YYYY-MM-DD (45 Days after submission)  |
+
+## Rationale for Migration
+* What are the use cases for the addon?
+* OSS usage, H5 Index, etc.
+
+## Historical Information
```
Comment (bhack) [inline sigs/addons/MIGRATION_RFC_TEMPLATE.md:15]: I don't know how it is easy to have a quick query but for stability we could ask how many bugs was handled in the last (x) month(s).

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+### Process
```
Comment (bhack) [inline sigs/addons/MIGRATION_TO_CORE.md:6]: I think this could be quite complicated for cases like `ImageProjectiveTransformV2`.
As we could have an upstream request where the sponsor is already in the TF core team I don't think this double pass is plausible for this specific case.

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+of our GitHub project.
+3. If it's agreed that a migration makes sense, an RFC needs to be written to discuss 
+the move with a larger community audience. 
+    * If the transition will impact tf-core and Keras then submit the RFC to 
+    [TensorFlow Community](https://github.com/tensorflow/community)
+    * Additions which only subclass Keras APIs should submit their migration proposals to 
+    [Keras Governance](https://github.com/keras-team/governance)
+    
+4. A sponsor from the TF/Keras team must agree to shepard the transition.
+   * If no sponsor is obtained after 45 days the RFC will be rejected and will remain
```
Comment (bhack) [inline sigs/addons/MIGRATION_TO_CORE.md:19]: I am ok to put a number here but I think that we need a realistic evaluation by the TF team. Is 45 days something realistic for the internal process?

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+   * If no sponsor is obtained after 45 days the RFC will be rejected and will remain 
+   as part of Addons.
+5. If a sponsor is obtained, and the RFC is approved, a pull request must move the 
+addon along with proper tests.
+6. After merging, the addition will be replaced with an alias to the core function 
+if possible. If an alias is not possible (e.g. large parameter changes), then a deprecation 
+warning will be added and will be removed from TFA after 2 releases. 
+
+
+### Criteria for Migration
```
Comment (bhack) [inline sigs/addons/MIGRATION_TO_CORE.md:28]: This ok for me on the SIG side. I just want to wait if TF team could put some quantitative metrics here.
Cause the sponsor define the bandwidth of the maintainership but having some quantitative metrics will make the process more transparent (if possible).

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+2. The SIG will evaluate the request and add it to the `Potential Candidates` section 
+of our GitHub project.
+3. If it's agreed that a migration makes sense, an RFC needs to be written to discuss 
+the move with a larger community audience. 
+    * If the transition will impact tf-core and Keras then submit the RFC to 
+    [TensorFlow Community](https://github.com/tensorflow/community)
+    * Additions which only subclass Keras APIs should submit their migration proposals to 
+    [Keras Governance](https://github.com/keras-team/governance)
+    
+4. A sponsor from the TF/Keras team must agree to shepard the transition.
```
Comment (alextp) [inline sigs/addons/MIGRATION_TO_CORE.md:18]: shepherd the transition and take over maintenance of the new API

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 6 above)
Comment (alextp) [inline sigs/addons/MIGRATION_TO_CORE.md:28]: ...and someone on the TensorFlow team agrees to take over maintenance of the API

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 1 above)
Comment (gabrieldemarmiesse) [reply sigs/addons/MIGRATION_TO_CORE.md:4]: It may be problematic since the Addons maintainers don't have write access to it, it may be possible that it's not updated often. If we have control over it, we can take over the responsability of updating it.

File: sigs/addons/MIGRATION_RFC_TEMPLATE.md (hunk 2 above)
Comment (gabrieldemarmiesse) [reply sigs/addons/MIGRATION_RFC_TEMPLATE.md:34]: We shouldn't alias it, but use the tf-core function in the body of our implementation. Otherwise we can't add a deprecation warning somewhere. As such it's ok if the parameters differs as long as it offers the same features/functionalities.

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 1 above)
Comment (bhack) [reply sigs/addons/MIGRATION_TO_CORE.md:4]: @gabrieldemarmiesse I agree with you cause we had not a really good historic records for this repo (see https://github.com/tensorflow/community/issues/226).
I think that the upstreaming/promotion process will enter soon or later in "submarine mode".
I suppose this will happen just after we have surpassed all the requirements/thresholds on the SIGs side and so as soon as the RFC could be formally opened here cause it is complaint with the bare minimal template thresholds/constrains (I hope also quantitative) that we are trying to define together in this review. 
We will have not so much control over the Kanban states cause the process at some point will be almost internal to the TF/core team (internal evaluations, internal activities looking for a sponsors etc.) so the ownership will be of the TF/core team and we can just suppose here that we could have some discrete "emergent" info about this phases of the process that could impact an update of a Kanban states here and/or  RFC labels handling. 
So I suppose we could manage a simplified Kanban on our side that is limited to our ownership of the process and so the last state before the final accepted/rejected could be just something like  "RFC submitted to community repo". 
From this point we need to track the process here cause we are totally dependent from discrete/sparse info emerging from the TF internal process and so if and when the TF team is able to release new info I think it could also handle the ownership of the state in Kanban and labels update.
We could think differently if the process will be totally public so we are on par about all the info related to the process or SIG members could have some different role here in the repo but as I understand with the current repo polcies IMHO this is not the case (see https://github.com/tensorflow/community/issues/227)

File: sigs/addons/MIGRATION_TO_CORE.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+1. Create an issue in TensorFlow Addons for a candidate that you think should be 
+migrated. 
+2. The SIG will evaluate the request and add it to the `Potential Candidates` section 
+of our GitHub project.
+3. If it's agreed that a migration makes sense, an RFC needs to be written to discuss 
+the move with a larger community audience. 
+    * If the transition will impact tf-core and Keras then submit the RFC to 
+    [TensorFlow Community](https://github.com/tensorflow/community)
+    * Additions which only subclass Keras APIs should submit their migration proposals to 
+    [Keras Governance](https://github.com/keras-team/governance)
```
Comment (bhack) [inline sigs/addons/MIGRATION_TO_CORE.md:16]: The new Keras website is online. I don't know if we need to duplicate/fork this template also in Keras governance and toadd something to the [Keras SIG new webpage](https://keras.io/governance/#governance-the-keras-sig). 
We are in the special case here where a SIG is an upstream of another SIG.

File: sigs/addons/MIGRATION_RFC_TEMPLATE.md (hunk 2 above)
Comment (alextp) [reply sigs/addons/MIGRATION_RFC_TEMPLATE.md:34]: I'd like to leave the core implementation the freedom to drop parameters that are not frequently used, and rename other parameters which might have better names.

Comment (seanpmorgan) [reply sigs/addons/MIGRATION_RFC_TEMPLATE.md]: Thanks, yeah even if it were a parameter match we'd still want to call the core function inside of our call with deprecation as mentioned. Updated the discussion points to keep them more general.

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 1 above)
Comment (seanpmorgan) [reply sigs/addons/MIGRATION_TO_CORE.md:4]: Agreed I think this is more sustainable if the Addons team is able to manage the project.

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 7 above)
Comment (seanpmorgan) [reply sigs/addons/MIGRATION_TO_CORE.md]: Updated.

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 6 above)
Comment (seanpmorgan) [reply sigs/addons/MIGRATION_TO_CORE.md]: Hmm so I think this is a bit of a chicken and egg problem. The sponsor will often be determined after the RFC has been created. I've renamed this as `Criteria for Migration RFC` PTAL and let me know if you think that makes more sense.

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 5 above)
Comment (bhack) [reply sigs/addons/MIGRATION_TO_CORE.md:20]: I think we could align to 1  month. See https://github.com/tensorflow/community/blob/master/governance/TF-RFCs.md#rfc-process

Comment (seanpmorgan) [reply sigs/addons/MIGRATION_TO_CORE.md:20]: Hmmm, would like to align here but 45 days ensures we overlap with a TFA monthly meeting and lets us get help from TF internal team members to help find sponsors. @theadactyl @ewilderj is there anyway we could make 45 days the standard amount of time for RFC sponsorship? I imagine this would be useful for multiple SIGs that have monthly meetings.

Comment (bhack) [reply sigs/addons/MIGRATION_TO_CORE.md:20]: Yes it is the same if we could extend the standard to 45 days is ok

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 8 above)
Comment (bhack) [reply sigs/addons/MIGRATION_TO_CORE.md:16]: @seanpmorgan We have other new guidelines for the Keras upstream: https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md

File: sigs/addons/MIGRATION_TO_CORE.md (hunk 4 above)
Comment (tanzhenyu) [reply sigs/addons/MIGRATION_TO_CORE.md:6]: This seems to be helping external contribution for the migration only, not internal contribution for the migration

Comment (bhack) [reply sigs/addons/MIGRATION_TO_CORE.md:6]: @tanzhenyu Do you mean that you want a different process for Google internals?

**Comment by seanpmorgan** [comment]
Per SIG discussion with TF team, we've determined that it is unfeasible for an external SIG to handle the migration process. In the future core TF and Keras member will handle the migration / deprecations:
https://github.com/tensorflow/addons/pull/2156
